lib/tools/system-calls.js

```diff
@@ -70,9 +71,127 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
     binaryLoc = stdout.trim();
     return binaryLoc;
   } catch(e) {
+    log.error(`Could not find ${binaryName} Please set the ANDROID_HOME ` +
+              `environment variable with the Android SDK root directory path.`);
     throw new Error(`Could not find ${binaryName} Please set the ANDROID_HOME ` +
                     `environment variable with the Android SDK root directory path.`);
   }
 };
 
+systemCallMethods.getConnectedDevices = async function () {
+  log.debug("Getting connected devices...");
+  try {
+    let {stdout} = await exec(this.adb.path, ['devices']);
+    // expecting adb devices to return output as
+    // List of devices attached
+    // emulator-5554	device
+    let startingIndex = stdout.indexOf("List of devices");
+    if (startingIndex === -1) {
+      log.error(`Unexpected output while trying to get devices. output was: ${stdout}`);
+      throw new Error(`Unexpected output while trying to get devices. output was: ${stdout}`);
+    } else {
+      // slicing ouput we care about.
+      stdout = stdout.slice(startingIndex);
+      let devices = [];
+      for (let line of stdout.split("\n")) {
+        if (line.trim() !== "" &&
+            line.indexOf("List of devices") === -1 &&
+            line.indexOf("* daemon") === -1 &&
+            line.indexOf("offline") === -1) {
+          let lineInfo = line.split("\t");
+          // state is either "device" or "offline", afaict
+          devices.push({udid: lineInfo[0], state: lineInfo[1]});
+        }
+      }
+      log.debug(`${devices.length} device(s) connected`);
+      return devices;
+    }
+  } catch (e) {
+    log.error(`Error while getting connected devices. Original error: ${e.message}`);
+    throw new Error(`Error while getting connected devices. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getDevicesWithRetry = async function (timeoutMs = 20000) {
```

Comment:
Catching up with the slack discussion. This should probably be refactored even before we start thinking about test, the error handling code is duplicated.